File: Data Structures & Algorithms/reconstruct-flight-path/submission-2.py
class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        graph = defaultdict(list)
        for a, b in tickets:
            # using heap to store the path
            heapq.heappush(graph[a], b)
        
        res = []

        def dfs(i):
            while graph[i]:
                dst = heapq.heappop(graph[i])
                dfs(dst)
            res.append(i)

        dfs("JFK")
        return res[::-1]

File: Data Structures & Algorithms/climbing-stairs/submission-2.cpp
class Solution {
private:
    int dfs(int n, vector<int>& memo) {
        if (n == 1 || n == 0) {
            return 1;
        }

        if (memo[n] != -1) {
            return memo[n];
        }

        return memo[n] = dfs(n - 1, memo) + dfs(n - 2, memo);
    }

public:
    int climbStairs(int n) {
        vector<int> memo(n + 1, -1);

        return dfs(n, memo);
    }
};
